[RFC] man.vim: parse page names containing dash or dot

noahfrederick opened the pull request:
Ensure that `'iskeyword'` contains `-` and `.`. Otherwise man.vim can fail to parse page names containing these characters:

```
nvim -c 'Man git-branch'
:Man limits.conf
```

Fixes #4165.

ZyX-I commented:
This does not look good. It is completely possible for &iskeyword to *not* contain alphanumeric characters. So do not use `-=`, `+=`, etc, use `set iskeyword={needed-value}`.

Though I would prefer to alter code which depends on a user setting to not depend on it.

ZyX-I commented:
Meaning that you need to replace `\k` atom in regexes below with `[collection]` with explicit characters.

noahfrederick commented:
@ZyX-I Agreed. I have re-pushed.

Note: I removed the leading `(*` in the original pattern for parsing `page`, since as far as I can tell, it is useless.

ZyX-I commented on runtime/autoload/man.vim:
```diff
@@ -118,15 +118,12 @@ endfunction
-    let page = substitute(a:str, '(*\(\k\+\).*', '\1', '')
-    let sect = substitute(a:str, '\(\k\+\)(\([^()]*\)).*', '\2', '')
+    let page = substitute(a:str, '\([-.[:alnum:]_]\+\).*', '\1', '')
+    let sect = substitute(a:str, '\([-.[:alnum:]_]\+\)(\([^()]*\)).*', '\2', '')
```
Can’t this be a single expression:

    let [page, sect] = matchlist(a:str, '\([-.[:alnum:]_]\+\)(\([^()]*\))')[1:2]

?

ZyX-I replied:
No,

    let [page, sect] = matchlist(a:str, '\v\C([-.[:alnum:]_]+)\(([^()]*)\)')[1:2]

: using “very magic” mode reduces the number of backslashes and `\C` should simply be a convention, regardless of whether or not it changes something in the regular expression (`match*` functions depend on &ignorecase option, establishing such convention should prevent the possible damage of unexpected option value, just like `#` after comparison operators which do string comparison).

ZyX-I replied:
I suggested such change earlier, but got confused by `(*` absent in the second regex. If these regular expressions are really fine, `matchlist()` could safely be used.

ZyX-I commented on runtime/autoload/man.vim:
```diff
@@ -118,15 +118,12 @@ endfunction
-    let sect = substitute(a:str, '\(\k\+\)(\([^()]*\)).*', '\2', '')
+    let page = substitute(a:str, '\([-.[:alnum:]_]\+\).*', '\1', '')
+    let sect = substitute(a:str, '\([-.[:alnum:]_]\+\)(\([^()]*\)).*', '\2', '')
     if sect == page || -1 == match(sect, '^[0-9 ]\+$')
```
With this suggestion `sect == page` should be replaced with `empty(sect)`.

ZyX-I replied:
And second part is confusing, one should write this as `sect =~# '^[0-9 ]\+$'`.

noahfrederick replied:
If rendered as

      let [page, sect] = matchlist(a:str, '\v\C([-.[:alnum:]_]+)%(\((\d+)\))?')[1:2]

the second part of the condition can be removed—`sect` is guaranteed to either be empty or contain only numeric characters.

This works for various inputs:

- `:Man man`
- `:Man man(1)`
- `:Man git-branch`
- `:Man foo.bar`

If that looks good, I will update the commit.

ZyX-I replied:
By the way, what with sections like `{number}p` and `{letter}`?

I have e.g. `man 3p printf` (“POSIX programmer’s manual”) and `man 3 printf` (“linux programmer’s manual”) describing C printf function, `man 1p printf` (“POSIX programmer’s manual”) and `man 1 printf` (“user commands”: this one is from coreutils) describing `/usr/bin/printf`. This is Gentoo, sys-apps/man-pages-posix package, I do not know whether such thing exists in your distribution.

I also have `n` section: this is from tcl, lapack, blas; all work. And not working `m` section from motif and `x` section from libcmis.

I actually suspect that section may be *any* string that can be a file name.

ZyX-I replied:
This makes test `if 0 + a:sect > 0` invalid.

noahfrederick replied:
> By the way, what with sections like {number}p and {letter}?

I wasn't aware of this, but yes, this should be accommodated. I'm only able to find examples in the forms

- `{number}`
- `{number}{letter}`
- `{letter}`

but I suppose there is no point in restricting the pattern to these options.

> This makes test if 0 + a:sect > 0 invalid.

The section name is assumed to be numeric in several places throughout the script, including `man#get_page()`. Changing this makes the single-argument version of the function ambiguous: the argument could either be a page name or section name (in which case the page is set to `expand('cword')`).

The function is referenced in `ftplugin/man.vim` with one argument:

```vim
nnoremap <silent> <buffer> <C-]> :call man#get_page(v:count)<CR>
" ...
nnoremap <silent> <buffer> K :call man#get_page(v:count)<CR>
```
where `[count]` can be used to specify the section number. I propose the following change:

```vim
nnoremap <silent> <buffer> <C-]> :call man#get_page(expand('<cword>'), v:count)<CR>
" ...
nnoremap <silent> <buffer> K :call man#get_page(expand('<cword>'), v:count)<CR>
```

And to always interpret a single argument as page name. I think this is easier to understand anyway ("it takes one required argument, the page name, and an optional second argument, the section name"). I realize `v:count` *must* be numeric, alleviating the ambiguity somewhat, but it shouldn't be assumed that `man#get_page()` will only be used in this context.

justinmk replied:
@noahfrederick :+1: So `man#get_page` will be less clever and will not do `expand('<cword>')` internally. Totally in favor of th.at

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -20,24 +22,13 @@ function man#get_page(...) abort
 
-  if 0 + sect > 0 && s:find_page(sect, page) == 0
-    let sect = ''
-  endif
```
Removing this was an oversight. I will restore it (with corrected condition).

noahfrederick commented:
I re-pushed with suggestions made by @ZyX-I. Namely, simplify parsing with `matchlist()` and allow for characters other than digits in `sect`. This included changing (simplifying) the way `man#get_page()` interprets its arguments.

ZyX-I commented on runtime/autoload/man.vim:
```diff
@@ -118,15 +111,11 @@ endfunction
       let sect = a:sect
     endif
   catch
-    let &l:iskeyword = save_isk
```
In the previous version &l:iskeyword did not get restored if everything is fine, now it is not modified. Do you think it was intentional and &iskeyword should be set to something specific after this function exits?

noahfrederick replied:
I suspect the original intent was to avoid such side effects, which should have restored `&l:iskeyword` in a `:finally` block rather than `:catch`. I see that `ftplugin/man.vim` sets it thus:

```vim
setlocal iskeyword+=\.,-,(,)
```

justinmk replied:
Yeah that wasn't intentional. It was either a bug I caused, or a bug from the original `man.vim`, which wasn't carefully written.

justinmk replied:
> I see that ftplugin/man.vim sets it thus:

That's intentional, so that `<cword>` works correctly while viewing a man page.

ZyX-I commented on runtime/autoload/man.vim:
```diff
@@ -20,25 +22,16 @@ function man#get_page(...) abort
-  endif
-
-  if s:find_page(sect, page) == 0
+  elseif s:find_page(sect, page) == 0
```
This change will call s:find_page only once, previously it did this for the second time with empty section when section was not empty, but page with this section was not found.

noahfrederick replied:
:+1: Not sure what I was thinking.

ZyX-I commented on runtime/autoload/man.vim:
```diff
@@ -142,8 +131,8 @@ endfunction
-  if where !~ "^/"
-    if matchstr(where, " [^ ]*$") !~ "^ /"
+  if where !~# "^/"
+    if matchstr(where, " [^ ]*$") !~# "^ /"
```
This is some crazy code: three (!) regexes used and two levels of `if`, while one regex can be used and zero levels:

```VimL
return (where =~# '^/\| /')
```

ZyX-I replied:
Or, better, zero regexes:

```VimL
return where[0] is# '/' || stridx(where, ' /') != -1
```

noahfrederick committed:
man.vim: fixes to argument handling and parsing

- Define a collection of legal characters when parsing page and section
  in `s:parse_page_and_section()` instead of relying on 'iskeyword',
  which is unreliable.
- Allow non-numeric section names (e.g., `3c`).
- Simplify argument handling in `man#get_page()` to accommodate
  non-numeric section names.

Fixes #4165.

noahfrederick committed:
man.vim: clean up regex/string comparisons

- Explicitly specify case sensitivity when comparing strings
- Clean up unnecessarily complex `if` statements

noahfrederick committed:
man.vim: prefer cmd-line to normal-mode commands

nhooyr commented:
Imo the behaviour of the builtin man page plugin sucks. I don't understand why it opens in a new tab by default, no autocomplete, and where the man page opens isn't really configurable. If we're gonna change it, I think we should instead opt for my plugin [neoman](https://github.com/nhooyr/neoman.vim). The code behind it sucks right now, but I'll improve it if its a welcomed change.

justinmk commented:
@nhooyr Definitely open to improving the default man plugin, but also don't want to review an entirely new plugin. Focused improvements would be good. Configuration is ok if it's tasteful.

nhooyr commented:
@justinmk its actually just an improvement over the default plugin. I just modified it, i didn't totally reinvent it. I'll create a PR soon with my proposed changes.

noahfrederick commented:
@justinmk @ZyX-I Sorry to leave this hanging. I addressed the comments. Also added 091ef2f, which amounts to a small improvement in style. Feel free to drop it if you think it's frivolous.

nhooyr commented:
nhooyr/neoman#15.

I rewrote most of the plugin.
- there are no complicated regexes, just a few of simple to follow if statements. 
- there is autocomplete
- it pretty prints errors on a single line (`echoerr` tells you the line number/function which should be mostly for the developer)
- the tagstack is more simple, its an actual stack.
- you can control the splitting
- you can control whether or not to to move to the most recent neoman window with the bang
- i updated the syntax file so that it recognizes section `n`, I also removed all the unnecessary clutter, it only applied to vim
- I split it into smaller functions
- I included all of the updates in this P.R
- If no section is given, it automatically finds the section to set the buffer name correctly.
- Other things I can't remember.

I have a few questions regarding `keywordprg`. `keywordprg` uses `<word> to pass the manpage to neoman. But most filetypes do not include `(,)` in `iskeyword` so `printf(3)` will instead jump to `printf(1)`. In neoman buffers `iskeyword` is set so that this does not happen, but I obviously can't set it globally.

I created a simple global mapping that calls `:Nman` without any arguments. If no arguments were given, I was using `<cword>` before but now I changed that to `cWORD` and then strip everything after the `)`, if any. Thus I don't need to set `iskeyword` anymore either.

However maybe I'll change it so that the global mapping uses `cWORD` but the mappings inside neoman buffers use `cword`. Thoughts?

justinmk commented:
@nhooyr Many of those changes sound good. Of course we want to keep the command `:Man`, not `:Nman`. So, why not make a PR?

nhooyr commented:
@justinmk Will do soon, I'm gonna just add syntax highlighting for other languages first.
